onnx/defs/data_type_utils.cc

```diff
@@ -0,0 +1,396 @@
+    GetTypeStrToProtoMap()[typeStr] = type_proto;
+  }
+  return &(GetTypeStrToProtoMap().find(typeStr)->first);
+}
+
+DataType DataTypeUtils::ToType(const std::string& type_str) {
+  TypeProto type;
+  FromString(type_str, type);
+  return ToType(type);
+}
+
+const TypeProto& DataTypeUtils::ToTypeProto(const DataType& data_type) {
+  std::lock_guard<std::mutex> lock(GetTypeStrLock());
+  auto it = GetTypeStrToProtoMap().find(*data_type);
+  assert(it != GetTypeStrToProtoMap().end());
+  return it->second;
+}
+
+std::string DataTypeUtils::ToString(
+    const TypeProto& type_proto,
+    const std::string& left,
+    const std::string& right) {
+  switch (type_proto.value_case()) {
+    case TypeProto::ValueCase::kTensorType: {
+      if (type_proto.tensor_type().has_shape() &&
+          type_proto.tensor_type().shape().dim_size() == 0) {
+        // Scalar case.
+        return left + ToDataTypeString(type_proto.tensor_type().elem_type()) +
+            right;
+      } else {
+        return left + "tensor(" +
+            ToDataTypeString(type_proto.tensor_type().elem_type()) + ")" + right;
+      }
+    }
+
+    case TypeProto::ValueCase::kSparseTensorType:
+      return left + "sparse(" +
+          ToDataTypeString(type_proto.sparse_tensor_type().elem_type()) + ")" +
+          right;
+
+    default:
+      assert(false);
+      return "";
+  }
+}
+
+std::string DataTypeUtils::ToDataTypeString(
+    const TensorProto::DataType& tensor_data_type) {
+  TypesWrapper& t = TypesWrapper::GetTypesWrapper();
+  auto iter = t.TensorDataTypeToTypeStr().find(tensor_data_type);
+  assert(t.TensorDataTypeToTypeStr().end() != iter);
+  return iter->second;
+}
+
+void DataTypeUtils::FromString(const std::string& type_str, TypeProto& type_proto) {
+  StringRange s(type_str);
+  type_proto.Clear();
+  if (s.LStrip("sparse")) {
+    s.ParensWhitespaceStrip();
+    TensorProto::DataType e;
+    FromDataTypeString(std::string(s.Data(), s.Size()), e);
+    type_proto.mutable_sparse_tensor_type()->set_elem_type(e);
+  } else if (s.LStrip("tensor")) {
+    s.ParensWhitespaceStrip();
+    TensorProto::DataType e;
+    FromDataTypeString(std::string(s.Data(), s.Size()), e);
+    type_proto.mutable_tensor_type()->set_elem_type(e);
+  } else {
+    // Scalar
+    TensorProto::DataType e;
+    FromDataTypeString(std::string(s.Data(), s.Size()), e);
+    TypeProto::TensorTypeProto* t = type_proto.mutable_tensor_type();
+    t->set_elem_type(e);
+    // Call mutable_shape() to initialize a shape with no dimension.
+    t->mutable_shape();
+  }
+}
+
+bool DataTypeUtils::IsValidDataTypeString(const std::string& type_str) {
+  TypesWrapper& t = TypesWrapper::GetTypesWrapper();
+  const auto& allowedSet = t.GetAllowedDataTypes();
+  return (allowedSet.find(type_str) != allowedSet.end());
+}
+
+void DataTypeUtils::FromDataTypeString(
+    const std::string& type_str,
+    TensorProto::DataType& tensor_data_type) {
+  assert(IsValidDataTypeString(type_str));
```

Comment:
I'd still make it something more strict than assert() here. What if somebody tries to call FromDataTypeString() with invalid string? I'd still prefer exceptions (or just crash but including release mode) here.

Another reason for exceptions - we can append the name of the schema later at the higher level

Reply:
Hmmm, OK, let me fix this in next PR. Basically, I'd do it in two points:
1. For operator registering, I'll keep using assert, which is ok, since it will introduce build break, I believe.
2. For other public calls (someone calls ToType), I'm thinking of add a "Status" class to return the result with detail error message. The reason that I still may not want to throw exception is it may be not good for windows to call a SDK which throws exception.